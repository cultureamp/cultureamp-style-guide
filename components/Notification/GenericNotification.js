// @flow
import * as React from 'react';
import Icon from '../Icon/Icon';
import successIcon from '../../icons/success.svg';
import exclamationIcon from '../../icons/exclamation.svg';
import infoIcon from '../../icons/information.svg';
import closeIcon from '../../icons/close.svg';
import styles from './Notification.module.scss';
import classnames from 'classnames';

type Props = {|
  type: 'positive' | 'warning' | 'negative' | 'info',
  style: 'global' | 'normal' | 'toast',
  persistent?: boolean,
  title?: string,
  children?: ?React.Node,
  autohide?: ?boolean,
  onHide?: ?() => void | any,
|};

type State = {
  isFading: boolean,
};

class GenericNotification extends React.Component<Props, State> {
  state = {
    isFading: true,
  };

  componentDidMount() {
    requestAnimationFrame(() => this.setState({ isFading: false }));

    if (this.props.style === 'toast' && this.props.autohide) {
      setTimeout(() => this.triggerHide(), 5000);
    }
  }

  render() {
    return (
      <div className={this.determineClassName()}>
        <div className={styles.icon}>
          <Icon icon={this.getIconType()} role="presentation" inheritSize />
        </div>
        <div className={styles.textContainer}>
          {this.props.title && (
            <span className={styles.title}>{this.props.title}</span>
          )}
          <span className={styles.textWrapper}>
            <span className={styles.text}>{this.props.children}</span>
          </span>
        </div>
        {!this.props.persistent && this.renderCancelButton()}
      </div>
    );
  }

  determineClassName(): string {
    return classnames(
      styles.notification,
      styles[this.props.type],
      styles[this.props.style],
      {
        [styles.fading]: this.state.isFading,
      }
    );
  }

  getIconType() {
    switch (this.props.type) {
      case 'positive':
        return successIcon;
      case 'negative':
        return exclamationIcon;
      case 'warning':
        return exclamationIcon;
      case 'info':
        return infoIcon;
      default:
        return infoIcon;
    }
  }

  renderCancelButton() {
    return (
      <span className={styles.cancel} onClick={() => this.triggerHide()}>
        <span className={styles.cancelInner}>
          <Icon
            icon={closeIcon}
            role="img"
            title="close notification"
            inheritSize
          />
        </span>
      </span>
    );
  }

  triggerHide() {
    requestAnimationFrame(() => this.setState({ isFading: true }));

    setTimeout(() => {
      this.props.onHide && this.props.onHide();
    }, 600);
  }
}

export default GenericNotification;
