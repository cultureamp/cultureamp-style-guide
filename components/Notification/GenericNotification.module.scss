@import 'styles';

.notification {
  @extend %ca-notification;
}

// Structure

.icon {
  @extend %ca-notification__icon;
}

.textContainer {
  @extend %ca-notification__text-container;
}

.title {
  @extend %ca-notification__title;
}

.textWrapper {
  @extend %ca-notification__textWrapper;
}

.text {
  @extend %ca-notification__text;
}

.cancel {
  @extend %ca-notification__cancel;
}

.cancelInner {
  @extend %ca-notification__cancel__inner;
}

.fading {
  @extend %ca-notification---fading;
}

// Notification Style

.inline {
  @extend %ca-notification---inline;
}

.toast {
  @extend %ca-notification---toast;
}

.global {
  @extend %ca-notification---global;
}

// Notification Type

.affirmative {
  @extend %ca-notification---affirmative;
}

.informative {
  @extend %ca-notification---informative;
}

.warning {
  @extend %ca-notification---warning;
}

.negative {
  @extend %ca-notification---negative;
}
