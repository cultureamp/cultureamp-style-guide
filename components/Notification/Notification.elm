module Notification.Notification
    exposing
        ( view
        , inline
        , toast
        , global
        , notificationType
        , automationId
        , Config
        , NotificationType(..)
        , NotificationState(..)
        , NotificationStage(..)
        , NotificationStateSetter
        , getAutomationId
        , notificationStage
        , subscriptions
        )

import Html exposing (Html, text, div, span, h6, p, button)
import Html.Attributes
import Html.Events as Events exposing (on, onWithOptions, defaultOptions)
import Json.Decode as Json
import CssModules exposing (css)
import Icon.Icon as Icon
import Icon.SvgAsset exposing (svgAsset)
import Platform.Sub
import AnimationFrame
import Time exposing (every, second)


{-| A notification component for Culture Amp projects.

The notification view requires two properties, a config, and a state.

The config begins by choosing one of our 3 variants (`inline`, `toast`, and
`global`). These can then be changed using the modifiers (`notificationType`,
`automationId`).

The state is a single value that describes the animation's enter and exit states.
It includes whether the notification requires Manual dismissal or is set
to Autohide, and what the current `NotificationStage` is.

Your page will need to use `Notification.subscriptions` to subscribe to events
so all state transitions animate correctly.

-}



-- CONFIG


type Config msg
    = Config (ConfigValue msg)


type alias ConfigValue msg =
    { variant : Variant
    , notificationType : NotificationType
    , title : Maybe String
    , content : List (Html msg)
    , persistent : Bool
    , automationId : Maybe String
    }


type Variant
    = Inline
    | Toast
    | Global


type NotificationType
    = Affirmative
    | Informative
    | Cautionary
    | Negative



-- STATE


type NotificationState
    = Manual NotificationStage
    | Autohide NotificationStage


type NotificationStage
    = Appearing
    | Visible
    | Disappearing Int
    | Hidden


type alias NotificationStateSetter msg =
    NotificationState -> msg



-- VIEW


view : Config msg -> NotificationState -> NotificationStateSetter msg -> Html msg
view (Config config) state onStateChange =
    let
        className =
            notificationClassName config state

        style =
            case notificationStage state of
                Disappearing height ->
                    [ Html.Attributes.style
                        [ ( "marginTop", toString (-height) ++ "px" )
                        ]
                    ]

                _ ->
                    []

        automationId =
            case config.automationId of
                Just id ->
                    [ Html.Attributes.attribute "data-automation-id" id ]

                Nothing ->
                    []

        -- When using Autohide, we use a "transitionstart" event to retrieve the height of the notification, so we can animate the exit correctly.
        onTransitionStart =
            case state of
                Autohide (Disappearing oldHeight) ->
                    [ on
                        "transitionstart"
                        (Json.at [ "target", "clientHeight" ]
                            Json.int
                            |> Json.andThen
                                (\height ->
                                    if height /= oldHeight then
                                        Json.succeed <| onStateChange <| Autohide (Disappearing height)
                                    else
                                        Json.fail "ignore"
                                )
                        )
                    ]

                _ ->
                    []

        -- Each of our CSS transitions finishes with "margin-top" being animated so that our height is collapsed.
        -- After this is finished, the state is set to "Hidden" and the notification is removed from the DOM.
        onTransitionEnd : List (Html.Attribute msg)
        onTransitionEnd =
            case notificationStage state of
                Disappearing _ ->
                    [ on
                        "transitionend"
                        (Json.field "propertyName" Json.string
                            |> Json.andThen
                                (\propertyName ->
                                    if propertyName == "margin-top" then
                                        Json.succeed <| onStateChange <| Manual Hidden
                                    else
                                        Json.fail "ignore"
                                )
                        )
                    ]

                _ ->
                    []
    in
        case notificationStage state of
            Hidden ->
                text ""

            _ ->
                div (className ++ style ++ automationId ++ onTransitionStart ++ onTransitionEnd)
                    [ viewIcon (Config config)
                    , div [ class .textContainer ]
                        [ viewTitle (Config config)
                        , p [ class .text ] config.content
                        ]
                    , viewCancelButton (Config config) state onStateChange
                    ]


notificationClassName : ConfigValue msg -> NotificationState -> List (Html.Attribute msg)
notificationClassName config state =
    [ classList
        [ ( .notification, True )
        , ( .inline, config.variant == Inline )
        , ( .toast, config.variant == Toast )
        , ( .global, config.variant == Global )
        , ( .affirmative, config.notificationType == Affirmative )
        , ( .informative, config.notificationType == Informative )
        , ( .cautionary, config.notificationType == Cautionary )
        , ( .negative, config.notificationType == Negative )
        , ( .hidden, (notificationStage state) /= Visible )
        ]
    ]


viewIcon : Config msg -> Html msg
viewIcon (Config { notificationType }) =
    let
        iconAsset =
            icon notificationType
    in
        div [ class .icon ]
            [ Icon.view
                (Icon.presentation |> Icon.inheritSize True)
                iconAsset
                |> Html.map never
            ]


icon : NotificationType -> Icon.SvgAsset.SvgAsset
icon type_ =
    case type_ of
        Affirmative ->
            svgAsset "cultureamp-style-guide/icons/success.svg"

        Cautionary ->
            svgAsset "cultureamp-style-guide/icons/exclamation.svg"

        Negative ->
            svgAsset "cultureamp-style-guide/icons/exclamation.svg"

        Informative ->
            svgAsset "cultureamp-style-guide/icons/information.svg"


viewTitle : Config msg -> Html msg
viewTitle (Config { title }) =
    case title of
        Just titleText ->
            h6 [ class .title ] [ text titleText ]

        Nothing ->
            text ""


viewCancelButton : Config msg -> NotificationState -> NotificationStateSetter msg -> Html msg
viewCancelButton (Config { persistent, variant }) state onStateChange =
    let
        -- With Toast Notifications, we only allow hiding the close button if the notification will autohide.
        hideCloseButton =
            case ( variant, state ) of
                ( Toast, Autohide _ ) ->
                    persistent

                ( Toast, _ ) ->
                    False

                ( _, _ ) ->
                    persistent

        onClickCancel =
            [ onWithOptions
                "click"
                { defaultOptions | preventDefault = True }
                (Json.at [ "target", "parentNode", "clientHeight" ] Json.int
                    |> Json.andThen (\height -> Json.succeed <| onStateChange <| Manual (Disappearing height))
                )
            ]
    in
        if hideCloseButton then
            text ""
        else
            button
                ([ class .cancel ]
                    ++ onClickCancel
                )
                [ span
                    [ class .cancelInner
                    ]
                    [ -- We are using a hidden span and Icon.presentation rather than the usual Icon.img to avoid this components API requiring a unique ID.
                      span [ class .cancelLabel ] [ text "close notification" ]
                    , Icon.view Icon.presentation
                        (svgAsset "cultureamp-style-guide/icons/close.svg")
                        |> Html.map never
                    ]
                ]


{ class, classList } =
    css "cultureamp-style-guide/components/Notification/components/GenericNotification.module.scss"
        { notification = ""
        , icon = ""
        , textContainer = ""
        , title = ""
        , text = ""
        , cancel = ""
        , cancelInner = ""
        , cancelLabel = ""
        , hidden = ""
        , inline = ""
        , toast = ""
        , global = ""
        , affirmative = ""
        , informative = ""
        , cautionary = ""
        , negative = ""
        }



-- VARIANTS


defaults : ConfigValue msg
defaults =
    { variant = Inline
    , notificationType = Informative
    , title = Nothing
    , content = []
    , persistent = False
    , automationId = Nothing
    }


inline : String -> List (Html msg) -> Bool -> Config msg
inline title content persistent =
    Config { defaults | title = Just title, content = content, persistent = persistent }


toast : String -> List (Html msg) -> Bool -> Config msg
toast title content hideCloseIcon =
    Config
        { defaults
            | variant = Toast
            , title = Just title
            , content = content
            , persistent = hideCloseIcon
        }


global : List (Html msg) -> Config msg
global content =
    Config { defaults | variant = Global, content = content }



-- MODIFIERS


notificationType : NotificationType -> Config msg -> Config msg
notificationType value (Config config) =
    Config { config | notificationType = value }


automationId : String -> Config msg -> Config msg
automationId value (Config config) =
    Config { config | automationId = Just value }



-- HELPERS


getAutomationId : Config msg -> Maybe String
getAutomationId config =
    -- In our Notification.Demo app we need to access the automationId as a way of tracking component states.
    case config of
        Config { automationId } ->
            automationId


notificationStage : NotificationState -> NotificationStage
notificationStage state =
    case state of
        Autohide stage ->
            stage

        Manual stage ->
            stage



-- SUBSCRIPTIONS


{-| Get a subscription that advances the state of all notifications on the page.

We use subscriptions to change from "Appearing" to "Visible" on the first
animation frame, and to hide "Autohide" toast notifications after 5 seconds.

To set up a subscription, provide a list of all notifications on the page, for
each providing the current state, and a message we can use to update the state.

-}
subscriptions : List ( NotificationState, NotificationState -> msg ) -> Sub msg
subscriptions allNotifications =
    Sub.batch
        (List.filterMap
            (\( state, setter ) ->
                case state of
                    Manual Appearing ->
                        Just <| AnimationFrame.times <| always <| setter <| Manual Visible

                    Autohide Appearing ->
                        Just <| AnimationFrame.times <| always <| setter <| Autohide Visible

                    Autohide Visible ->
                        -- Note: we do not know the height of the notification here, so cannot animate margin-top.
                        -- We have a "transitionstart" event listener that will read the clientHeight and correct this value.
                        Just <| every (5 * second) <| always <| setter <| Autohide (Disappearing 0)

                    _ ->
                        Nothing
            )
            allNotifications
        )
