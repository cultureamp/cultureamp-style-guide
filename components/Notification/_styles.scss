@import '../../styles/color';
@import '../../styles/type';
@import '../../styles/border';

$caNotification-toast-fade-time: 0.5s;

// Structure

%ca-notification {
  display: flex;
  position: relative;

  a {
    color: $ca-link-text-color;
    text-decoration: underline;
    cursor: pointer;

    &:hover {
      color: $ca-link-hover-text-color;
      text-decoration: none;
    }
  }

  strong {
    font-weight: $ca-weight-semibold;
  }
}

%ca-notification__icon {
  align-items: flex-start;
  width: $ca-grid;
  height: $ca-grid;
  position: relative;
  margin-top: $ca-grid / 4;
}

%ca-notification__text-container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  flex: 1 0 0;
  margin-left: $ca-grid / 2;
}

%ca-notification__title {
  @include ca-type-ideal-heading;
  display: block;
  padding-right: $ca-grid / 2;
}

%ca-notification__textWrapper {
  @include ca-type-block;
}

%ca-notification__text {
  display: block;
}

%ca-notification__cancel {
  color: add-tint($ca-palette-lapis, 40%);
  position: relative;
  display: flex;
  align-items: center;
  justify-content: flex-end;

  &:hover {
    color: add-tint($ca-palette-lapis, 20%);
    cursor: pointer;
  }
}

%ca-notification__cancel__inner {
  $size: 20px;
  height: $size;
  width: $size;
  margin-right: 10px;
}

%ca-notification---fading {
  opacity: 0;
  transition: opacity $caNotification-toast-fade-time linear,
    transform ($caNotification-toast-fade-time * 0.8) ease-in;

  %ca-notification__cancel:hover {
    cursor: auto;
  }
}

// Notification Style

%ca-notification---normal,
%ca-notification---toast {
  $borderWidth: 1px;
  $verticalPadding: calc(#{$ca-grid / 4} - #{$borderWidth});
  $horizontalPadding: 20px;
  padding: $verticalPadding $horizontalPadding;

  margin-bottom: $ca-grid;
  border-width: $borderWidth;
  border-style: solid;
  border-radius: $ca-border-radius;

  %ca-notification__text {
    @include ca-type-ideal-notification;
    // When the title and text are on different lines this ensures the baselines are exactly 1 grid unit apart.
    // When they are both on a single line it ensures both are 1 grid unit tall.
    margin-top: $ca-grid / 4;
  }

  %ca-notification__cancel {
    width: $ca-grid * 2;
    height: $ca-grid * 2;
    // The negative margins are so that our clickable area will overlap the notification padding
    margin-right: -20px;
    margin-top: -$ca-grid / 4;
    margin-bottom: -$ca-grid / 4;
  }
}

%ca-notification---toast {
  width: 25rem;
  box-shadow: 0 3px 7px add-alpha($ca-palette-ink, 9%);
  transition: opacity $caNotification-toast-fade-time linear,
    transform ($caNotification-toast-fade-time * 0.8) ease-out;
  will-change: transform;

  &%ca-notification---fading {
    transform: translateX(50%);
  }
}

%ca-notification---global {
  padding: $ca-grid / 2 20px;
  width: 100%;
  border-width: 0;
  border-radius: 0;
  margin-bottom: 0;

  %ca-notification__text {
    @include ca-type-ideal-body;
  }

  %ca-notification__cancel {
    width: $ca-grid * 2.5;
    height: $ca-grid * 2.5;
    // The negative margins are so that our clickable area will overlap the notification padding
    margin-right: -20px;
    margin-top: -$ca-grid / 2;
    margin-bottom: -$ca-grid / 2;
  }
}

// Notification Type

%ca-notification---positive {
  background: add-tint($ca-palette-seedling, 90%);
  border-color: add-tint($ca-palette-seedling, 60%);
  color: $ca-palette-lapis;

  %ca-notification__icon {
    color: add-tint($ca-palette-seedling, 40%);
  }
}

%ca-notification---warning {
  background: add-tint($ca-palette-yuzu, 95%);
  border-color: add-tint($ca-palette-yuzu, 60%);
  color: $ca-palette-lapis;

  %ca-notification__icon {
    color: add-tint($ca-palette-yuzu, 20%);
  }
}

%ca-notification---negative {
  background: add-tint($ca-palette-coral, 95%);
  border-color: add-tint($ca-palette-coral, 60%);
  color: $ca-palette-lapis;

  %ca-notification__icon {
    color: add-tint($ca-palette-coral, 20%);
  }
}

%ca-notification---info {
  background: add-tint($ca-palette-ocean, 90%);
  border-color: add-tint($ca-palette-ocean, 60%);
  color: $ca-palette-lapis;

  %ca-notification__icon {
    color: add-tint($ca-palette-ocean, 30%);
  }
}
