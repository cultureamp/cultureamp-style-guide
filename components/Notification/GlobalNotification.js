// @flow
import * as React from 'react';
import GenericNotification from './components/GenericNotification';
import type { NotificationType } from './components/GenericNotification';

type Props = {|
  type: NotificationType,
  children: React.Node,
  onHide?: () => void,
|};

const GlobalNotification = (props: Props) => (
  <GenericNotification style="global" persistent={false} {...props} />
);

export default GlobalNotification;
