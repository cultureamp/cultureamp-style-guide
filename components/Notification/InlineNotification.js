// @flow
import * as React from 'react';
import GenericNotification from './GenericNotification';

type Props = {|
  type: 'affirmative' | 'informative' | 'warning' | 'negative',
  title: string,
  children: React.Node,
  persistent?: boolean,
  onHide?: () => void,
|};

const InlineNotification = (props: Props) => (
  <GenericNotification style="normal" {...props} />
);

export default InlineNotification;
