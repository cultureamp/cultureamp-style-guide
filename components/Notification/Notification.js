// @flow
import * as React from 'react';
import GenericNotification from './GenericNotification';

type Props = {|
  type: 'positive' | 'warning' | 'negative' | 'info',
  title: string,
  children: React.Node,
  persistent?: boolean,
  onHide?: () => void,
|};

const Notification = (props: Props) => (
  <GenericNotification style="normal" {...props} />
);

export default Notification;
