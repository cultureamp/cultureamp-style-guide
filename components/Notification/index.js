// @flow
import GenericNotification from './GenericNotification.js';
import GlobalNotification from './GlobalNotification.js';
import Notification from './Notification.js';
import ToastNotification from './ToastNotification.js';

export {
  GenericNotification,
  GlobalNotification,
  Notification,
  ToastNotification,
};

export default Notification;
