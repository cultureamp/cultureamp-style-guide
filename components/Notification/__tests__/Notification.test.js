import React from 'react';
import { mount } from 'enzyme';
import Notification from '../Notification';

test('The basic notification renders correctly', () => {
  const notification = mount(
    <Notification type="warning" title="Warning">
      Something has gone wrong
    </Notification>
  );
  expect(notification.find('.cancel')).toHaveLength(1);
  expect(notification).toMatchSnapshot();
});

test('Persistent versions of the notifications work', () => {
  const notification = mount(
    <Notification type="negative" title="Error" persistent>
      Something has gone wrong
    </Notification>
  );
  expect(notification.find('.cancel')).toHaveLength(0);
  expect(notification).toMatchSnapshot();
});
