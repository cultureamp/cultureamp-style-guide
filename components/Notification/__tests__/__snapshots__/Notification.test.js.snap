// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Persistent versions of the notifications work 1`] = `
<Notification
  persistent={true}
  title="Error"
  type="negative"
>
  <GenericNotification
    persistent={true}
    style="normal"
    title="Error"
    type="negative"
  >
    <div
      className="notification negative normal fading"
    >
      <div
        className="icon"
      >
        <Icon
          icon={null}
          inheritSize={true}
          role="presentation"
        >
          <svg
            aria-hidden={true}
            className="icon inheritSize"
            focusable="false"
            role="presentation"
            viewBox="viewBox"
          >
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </div>
      <div
        className="textContainer"
      >
        <span
          className="title"
        >
          Error
        </span>
        <span
          className="textWrapper"
        >
          <span
            className="text"
          >
            Something has gone wrong
          </span>
        </span>
      </div>
    </div>
  </GenericNotification>
</Notification>
`;

exports[`The basic notification renders correctly 1`] = `
<Notification
  title="Warning"
  type="warning"
>
  <GenericNotification
    style="normal"
    title="Warning"
    type="warning"
  >
    <div
      className="notification warning normal fading"
    >
      <div
        className="icon"
      >
        <Icon
          icon={null}
          inheritSize={true}
          role="presentation"
        >
          <svg
            aria-hidden={true}
            className="icon inheritSize"
            focusable="false"
            role="presentation"
            viewBox="viewBox"
          >
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </div>
      <div
        className="textContainer"
      >
        <span
          className="title"
        >
          Warning
        </span>
        <span
          className="textWrapper"
        >
          <span
            className="text"
          >
            Something has gone wrong
          </span>
        </span>
      </div>
      <span
        className="cancel"
        onClick={[Function]}
      >
        <span
          className="cancelInner"
        >
          <Icon
            icon={null}
            inheritSize={true}
            role="img"
            title="close notification"
          >
            <svg
              aria-labelledby="id-2-title"
              className="icon inheritSize"
              focusable="false"
              role="img"
              viewBox="viewBox"
            >
              <title
                id="id-2-title"
              >
                close notification
              </title>
              <use
                xlinkHref="#id"
              />
            </svg>
          </Icon>
        </span>
      </span>
    </div>
  </GenericNotification>
</Notification>
`;
